### PR TITLE
Identity link with sso csharp

### PR DESCRIPTION
**Pull request by NikitaJain-MSFT**

### CODE REVIEW

File: samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Controllers/TabController.cs [hunk 1]
```diff
@@ -34,7 +35,7 @@ public class TabController : Controller
 
         // Get user access token.
         [HttpPost]
-        [Route("GetUserAccessToken")]
+        [Route("GetProfileOnBehalfOf")]
         public async Task<JsonResult> GetUserAccessToken(string accessToken, string tid, string userName)
```
Comment (v-chetsh) [inline samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Controllers/TabController.cs:39]: We are not getting GetUserAccessToken.
Rename it properly like GetUserDetails

File: samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Controllers/TabController.cs [hunk 2]
```diff
@@ -48,7 +49,8 @@ public async Task<JsonResult> GetUserAccessToken(string accessToken, string tid,
 
                 var photo = await client.GetPhotoAsync();
 
-                var userInfo = new UserData() {
```
Comment (v-chetsh) [inline samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Controllers/TabController.cs:51]: new User()

Comment (v-chetsh) [inline samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Controllers/TabController.cs:51]: var userInfo => var userDetail

File: samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Controllers/TabController.cs [hunk 3]
```diff
@@ -132,7 +134,18 @@ public JsonResult GetUserMapData(string userName)
 
             if (index != -1)
             {
-                var userInfo = UserMapData[index];
+                var userInfo = new UserMapData()
```
Comment (v-chetsh) [inline samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Controllers/TabController.cs:137]: new UserMapData() => new UserMapping()

File: samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Controllers/TabController.cs [hunk 4]
```diff
@@ -132,7 +134,18 @@ public JsonResult GetUserMapData(string userName)
...
+                    isAadSignedIn = UserMapData[index].isAadSignedIn,
+                    AadToken = "",
+                    FacebookId = UserMapData[index].FacebookId,
+                    FacebookToken = "",
+                    isFacebookSignedIn = UserMapData[index].isFacebookSignedIn,
+                    GoogleId = UserMapData[index].GoogleId,
+                    GoogleToken = "",
+                    isGoogleSignedIn = UserMapData[index].isGoogleSignedIn,
+                };
                 var jsonString = JsonConvert.SerializeObject(userInfo);
```
Comment (v-chetsh) [inline samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Controllers/TabController.cs:149]: Rename jsonString to userDetailJson

File: samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Controllers/TabController.cs [hunk 5]
```diff
@@ -145,7 +158,7 @@ public JsonResult GetUserMapData(string userName)
 
         // Get facebook profile of user.
         [HttpPost]
-        [Route("getFbAccessToken")]
+        [Route("getFbDetailsOauth")]
```
Comment (v-chetsh) [inline samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Controllers/TabController.cs:161]: Rename getFbDetailsOauth => getFbUserDetails

File: samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Controllers/TabController.cs [hunk 6]
```diff
@@ -232,7 +247,7 @@ public JsonResult DisconnectFb(string userName)
 
         // Get google profile of user.
         [HttpPost]
-        [Route("getGoogleAccessToken")]
+        [Route("getGoogleDetailsOauth")]
```
Comment (v-chetsh) [inline samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Controllers/TabController.cs:250]: Rename it to getGoogleAccountDetails or getGoogleUserDetails

File: samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Controllers/TabController.cs [hunk 7]
```diff
@@ -247,9 +262,9 @@ public async Task<JsonResult> GetGoogleAccessToken(string accessToken, string us
             {
                 try
                 {
-                responseBody = await response.Content.ReadAsStringAsync();
-                var token = JsonConvert.DeserializeObject<dynamic>(responseBody).access_token;
-                var client2 = new HttpClient();
+                    responseBody = await response.Content.ReadAsStringAsync();
+                    var token = JsonConvert.DeserializeObject<dynamic>(responseBody).access_token;
```
Comment (v-chetsh) [inline samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Controllers/TabController.cs:266]: You can rename token as accessToken or idToken

File: samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Controllers/TabController.cs [hunk 8]
```diff
@@ -314,11 +330,13 @@ public async Task<JsonResult> GetGoogleAccessToken(string accessToken, string us
         // Get google profile of existing mapped user.
         [HttpPost]
         [Route("getGoogleDetails")]
-        public async Task<JsonResult> GetGoogleDetails(string token)
+        public async Task<JsonResult> GetGoogleDetails(string userName)
         {
             var client2 = new HttpClient();
             client2.DefaultRequestHeaders.Accept.Clear();
```
Comment (v-chetsh) [inline samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Controllers/TabController.cs:336]: What is client2?
Can you rename it properly

File: samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Pages/tab.cshtml [hunk 9]
```diff
@@ -138,14 +134,15 @@
                     }
                     else {
                         const userDetails = JSON.parse(responseJson);
+                        console.log(userDetails);
                         if (userDetails.isAadSignedIn) {
                             if (userDetails.isGoogleSignedIn) {
                                 is_google_signed_in = true;
-                                setTimeout(() => { getGoogleDetails(userDetails.GoogleToken) }, 3000);
+                                setTimeout(() => { getGoogleDetails() }, 3000);
```
Comment (v-chetsh) [inline samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Pages/tab.cshtml:141]: getGoogleDetails rename it as getGoogleAccountDetails or getGoogleAccountUserDetails

File: samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Pages/tab.cshtml [hunk 10]
```diff
@@ -181,29 +179,27 @@
...
         }
-
         // Get server side token and user profile.
         function getServerSideToken(clientSideToken) {
             return new Promise((resolve, reject) => {
                 microsoftTeams.getContext((context) => {
                     $.ajax({
                         type: 'POST',
-                        url: '/GetUserAccessToken',
+                        url: '/GetProfileOnBehalfOf',
```
Comment (v-chetsh) [inline samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Pages/tab.cshtml:202]: What profile we are getting here?
I think we are getting User details so can you please rename it accordingly.

File: samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Pages/tab.cshtml [hunk 11]
```diff
@@ -321,15 +312,13 @@
...
-
         // Get face book profile of user.
         function getServerSideTokenFb(clientSideToken) {
             return new Promise((resolve, reject) => {
                 microsoftTeams.getContext((context) => {
-
                     $.ajax({
                         type: 'POST',
-                        url: '/getFbAccessToken',
+                        url: '/getFbDetailsOauth',
```
Comment (v-chetsh) [inline samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Pages/tab.cshtml:321]: getFbDetailsOauth rename it as getFbAccountDetails or getFbUserAccountDetails

File: samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Pages/tab.cshtml [hunk 12]
```diff
@@ -381,15 +368,13 @@
...
         // Get face book profile of user.
         function getServerSideTokenGoogle(clientSideToken) {
             return new Promise((resolve, reject) => {
                 microsoftTeams.getContext((context) => {
-
                     $.ajax({
                         type: 'POST',
-                        url: '/getGoogleAccessToken',
+                        url: '/getGoogleDetailsOauth',
                         dataType: 'json',
```
Comment (v-chetsh) [inline samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Pages/tab.cshtml:378]: Are we getting OAuth details or User details?
If User details, please rename it.

File: samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Pages/tab.cshtml [hunk 13]
```diff
@@ -412,15 +397,14 @@
...
-            console.log("da" + token);
+        function getFbDetails() {
+            microsoftTeams.getContext((context) => {
             $.ajax({
                 type: 'POST',
                 url: '/getFbDetails',
                 dataType: 'json',
                 data: {
-                    'token': token
+                    'userName': context.upn
```
Comment (v-chetsh) [inline samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Pages/tab.cshtml:407]: It is not user name, It's UserPrinciplename

File: samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Pages/tab.cshtml [hunk 14]
```diff
@@ -433,16 +417,17 @@
                     console.log("textStatus: " + textStatus + ", errorThrown:" + errorThrown);
                     fbAuth();
                 }
-            })
+             })
+          })
         }
-
-        function getGoogleDetails(token) {
+        function getGoogleDetails() {
```
Comment (v-chetsh) [inline samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Pages/tab.cshtml:423]: getGoogleAccountDetails or getGoogleUserAccountDetails

File: samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Pages/tab.cshtml [hunk 15]
```diff
@@ -433,16 +417,17 @@
...
+            microsoftTeams.getContext(function (context) {
             $.ajax({
                 type: 'POST',
                 url: '/getGoogleDetails',
                 dataType: 'json',
                 data: {
-                    'token': token
```
Comment (v-chetsh) [inline samples/app-identity-linking-with-sso/csharp/IdentityLinkingWithSSO/Pages/tab.cshtml:445]: It is not user name, It's UserPrinciplename

File: samples/app-identity-linking-with-sso/csharp/README.md [hunk 16]
```diff
@@ -22,10 +22,16 @@ Bot Mapping
 
 ![Bot](IdentityLinkingWithSSO/Images/BotLinkingOptionCard.png)
 
+- Click on `Connect to facebook` or `Connect to google` button you will get a sign in card. Click on sign in button and provide your facebook or google credentials.
```
Comment (v-chetsh) [inline samples/app-identity-linking-with-sso/csharp/README.md:25]: This is just explaining the steps to click on buttons.
We need to add the explanation of functionality like how the mapping or identity linking is working with different auth.

File: samples/app-identity-linking-with-sso/csharp/README.md [hunk 17]
```diff
@@ -43,6 +49,8 @@ Messaging Extention mapping
 - ME Link unfurling
 
   ![ME Link unfurling](IdentityLinkingWithSSO/Images/MELinkUnfurlingCard.png)
+
+  - On Click of `Disconnect from facebook` or `Disconnect from google`. User's facebook/google data will be removed from the data base.
```
Comment (v-chetsh) [inline samples/app-identity-linking-with-sso/csharp/README.md:53]: Database
Also, are we using any database or just saving it in memory.
So we can add a extensibility note like, you can use any memory like DB or in memory or in-file.

**Review by v-chetsh — COMMENTED** [review]
Reviewed and provided my comments.

**Review by v-chetsh — APPROVED** [review]
Changes looks fine, approving it.

**Review by Wajeed-msft — APPROVED** [review]

**Pull request merged**